phat/contracts/action_offchain_rollup/lib.rs

```diff
@@ -0,0 +1,357 @@
+                .ok_or(Error::NoRequestInQueue)?;
+            // Decode the queue data by ethabi (u256, bytes)
+            let Ok(decoded) = ethabi::decode(&[ParamType::Uint(32), ParamType::Bytes], &raw_req) else {
+                return Ok(LensResponse::Error(None, Error::FailedToDecode))
+            };
+            let [Token::Uint(rid), Token::Bytes(profile_id)] = decoded.as_slice() else {
+                return Err(Error::FailedToDecode);
+            };
+            let profile_id = String::from_utf8_lossy(&profile_id);
+            pink::info!("Request received for profile {profile_id}");
+
+            // Get the Lens stats and respond as a rollup action.
+            let result = self.fetch_lens_api_stats(String::from(profile_id));
+            match result {
+                Ok(res) => {
+                    // Respond
+                    pink::info!("Precessed result: {res}");
+                    Ok(LensResponse::Response(*rid, res))
+                }
+                // Error when fetching the stats. Could be
+                Err(Error::FailedToDecode) => {
+                    Ok(LensResponse::Error(Some(*rid), Error::FailedToDecode))
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        /// Returns BadOrigin error if the caller is not the owner
+        fn ensure_owner(&self) -> Result<()> {
+            if self.env().caller() == self.owner {
+                Ok(())
+            } else {
+                Err(Error::BadOrigin)
+            }
+        }
+
+        /// Returns the config reference or raise the error `NotConfigured`
+        fn ensure_configured(&self) -> Result<&Config> {
+            self.config.as_ref().ok_or(Error::NotConfigured)
+        }
+
+        /// Returns the client config reference or raise the error `NotConfigured`
+        fn ensure_client_configured(&self) -> Result<&ClientConfig> {
+            self.client_config
+                .as_ref()
+                .ok_or(Error::ClientNotConfigured)
+        }
+    }
+
+    enum LensResponse {
+        Response(U256, u128),
+        Error(Option<U256>, Error),
+    }
+
+    fn connect(client_config: &ClientConfig) -> Result<EvmRollupClient> {
+        let anchor_addr: H160 = client_config.anchor_addr.into();
+        EvmRollupClient::new(&client_config.rpc, anchor_addr)
+            .log_err("failed to create rollup client")
+            .or(Err(Error::FailedToCreateClient))
+    }
+
+    fn maybe_submit_tx(client: EvmRollupClient, config: &Config) -> Result<Option<Vec<u8>>> {
+        use pink_web3::keys::pink::KeyPair;
+        let maybe_submittable = client
+            .commit()
+            .log_err("failed to commit")
+            .or(Err(Error::FailedToCommitTx))?;
+        if let Some(submittable) = maybe_submittable {
+            let attest_pair = KeyPair::from(config.attest_key);
+            let tx_id = submittable
+                .submit_meta_tx(&attest_pair, config.account_contract)
```

Comment:
Refactored in this repo. But I am not sure whether this should be applied back to the official repo since the data to be signed by AccountContract is in the specific format of `pink_web3::types::TransactionRequest`.

Reply:
Good to merge back to the upstream. Ideally you can define a trait `Signer`, and implement the trait in your contract code. Then you can pass the signer to `submit_meta_tx()`.